### ISSUE TITLE
CompoundModel slicing does not permit all valid slices

### ISSUE THREAD

**Issue by chris-simpson** [astropy/astropy#10074]
### Description

CompoundModel slicing depends on the tree structure, and this structure does not permit all valid slicing operations. In particular, for long chains of models combined with |, the valid slices depend on the order in which this chain was constructed. So if I construct `m = A | B | C`, where A, B, and C are models, then I can slice to get `m[:2] = A | B`, but not `m[1:] = B | C`.

I can create `m = A | (B | C)`, which allows me to slice `m[1:] = B | C` but now I can't do `m[:2] = A | B`. This might be reasonable behaviour, since I have explicitly connected B and C and suggested they're inseparable, but the first example _should_ allow all slices. The problem, of course, is that it's constructed like `m = (A | B) | C` because every operator produces a `CompoundModel` that creates a tree node, but in reality there are really three leaves springing from a single top node.

## Expected behavior

I expected to obtain a valid slice that combined the last two submodels.

### Actual behavior

`IndexError: No appropriate subtree matches slice`

## Steps to Reproduce

# Put your Python code snippet here.
```
from astropy.modeling.models import Shift

m1 = Shift(1)
m2 = Shift(2)
m3 = Shift(3)

m123 = m1 | m2 | m3

m23 = m123[1:]
```

## System Details

Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.0
Scipy 1.4.1

**Comment by pllim** [comment]
cc @perrygreenfield and @nden

**Comment by perrygreenfield** [comment]
Yes, this is a consequence of the redesign of the compound models. Fixing it is not trivial, but it should be done at some time, but it won't be soon. For many cases, grouping is the suggested solution, but that won't solve all problems. Do you have a use case that makes this capability essential?

**Comment by chris-simpson** [comment]
I see the difficulty in creating a structure that allows all valid slicing options so I wasn't expecting a quick fix. I encountered it while manipulating gWCS transformations, but looking at the source code and understanding the cause of the problem has allowed me to work around it for what I'm trying to do (recasting a gWCS as FITS WCS keywords).
